src/lib/core/utils/data-element-constraints.ts

```diff
@@ -0,0 +1,123 @@
+import { mapStructure } from '@veupathdb/wdk-client/lib/Utils/TreeUtils';
+import { isEmpty, union } from 'lodash';
+import { StudyEntity } from '../types/study';
+import { Variable } from '../types/variable';
+import { DataElementConstraint } from '../types/visualization';
+import { findEntityAndVariable } from './study-metadata';
+
+export function filterVariablesByConstraint(
+  rootEntity: StudyEntity,
+  constraint?: DataElementConstraint
+): StudyEntity {
+  if (
+    constraint == null ||
+    (constraint.allowedShapes == null && constraint.allowedTypes == null)
+  )
+    return rootEntity;
+  return mapStructure(
+    (entity, children) => ({
+      ...entity,
+      variables: entity.variables.filter(
+        (variable) =>
+          variable.dataShape == null ||
+          variable.type === 'category' ||
+          ((constraint.allowedShapes == null ||
+            constraint.allowedShapes.includes(variable.dataShape)) &&
+            (constraint.allowedTypes == null ||
+              constraint.allowedTypes.includes(variable.type)))
+      ),
+      children,
+    }),
+    (e) => e.children ?? [],
+    rootEntity
+  );
+}
+
+export type ValueByInputName = Partial<Record<string, Variable>>;
```

Comment:
Could you explain the use of `Partial<>` here for me please?  Does it allow a record with no entries?

Reply:
`Partial<T>` means that some values in `T` may not be defined. With `Record<string, Variable>` on its own, all values are considered to be `Variable`. Adding `Partial` means that values can be `Variable` or `undefined`.